### PR TITLE
Acquire lock on singleton to handle parallel pod deployments

### PR DESCRIPTION
**Pull request by mad-anne**
Hello (again)! 😄
I tried to move further using your package and I've noticed the following issue along with my team.

## Issue

When deploying to multiple pods, we may encounter a situation when more than one pod runs the `handle_migration_zero_reset` command at the same time. It may cause issues on deployment.

## Solution

The `select_for_update` QuerySet method could be used to create a lock on a row with singleton and ensure a single access at a time.

It required changing a Manager class to a QuerySet class for chaining QS methods. I also modified the `fetch_singleton` method to use a single query on the locked row instead of two queries.

### Note

The test DB in the project runs on the SQLite that doesn't work with `select_for_update`. That's why I set the `number_of_pods = 1` in the test. When set to 3, it fails. But I've also tested the package on the PostgreSQL project and the test passed with `number_of_pods = 3` thanks to the `select_for_update`. Changing a DB engine to the one supporting `SELECT ... FOR UPDATE` could help here.

### CODE REVIEW

File: django_migration_zero/managers.py [hunk 1]
```diff
@@ -1,35 +1,26 @@
...
 class MigrationZeroConfigurationQuerySet(models.QuerySet):
-    pass
-
-
-class MigrationZeroConfigurationManager(models.Manager):
     def fetch_singleton(self) -> None:
         logger = get_logger()
         try:
-            number_records = self.count()
+            config_singleton = self.get()
```
Comment (GitRon) [inline django_migration_zero/managers.py:12]: praise: i like that 👍

File: django_migration_zero/models.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ class MigrationZeroConfiguration(models.Model):
     )
     migration_date = models.DateField(_("Migration date"), null=True, blank=True)
 
-    objects = MigrationZeroConfigurationManager()
+    objects = MigrationZeroConfigurationQuerySet.as_manager()
```
Comment (GitRon) [inline django_migration_zero/models.py:17]: question: why are you changing the manger setup? I don't see any reason why to do so in the other changes 🙂

File: django_migration_zero/services/deployment.py [hunk 3]
```diff
@@ -20,11 +21,12 @@ def __init__(self):
 
         self.logger = get_logger()
 
+    @transaction.atomic
     def process(self):
         self.logger.info("Starting migration zero database adjustments...")
 
         # Fetch configuration singleton from database
-        config_singleton = MigrationZeroConfiguration.objects.fetch_singleton()
+        config_singleton = MigrationZeroConfiguration.objects.select_for_update().fetch_singleton()
```
Comment (GitRon) [inline django_migration_zero/services/deployment.py:29]: praise: nice catch, together with atomic!

File: tests/services/test_deployment.py [hunk 4]
```diff
@@ -81,3 +83,56 @@ def test_process_validate_all_commands_are_executed(self, mocked_call_command):
...
+        config, _ = MigrationZeroConfiguration.objects.update_or_create(
+            defaults={
+                "migration_imminent": True,
+                "migration_date": timezone.now().date(),
+            }
+        )
+
+        # Testable
+        number_of_pods = 1
+        threads = [Thread(target=DatabasePreparationService().process) for _ in range(number_of_pods)]
```
Comment (GitRon) [inline tests/services/test_deployment.py:104]: praise: uff, really hard to write a test for that kind of setup. 👍

**Review by GitRon — COMMENTED** [review]
I have one question left, rest seems great!

File: django_migration_zero/models.py (hunk 2 above)
Comment (mad-anne) [reply django_migration_zero/models.py:17]: I changed it to a QuerySet class so it's possible to use chained methods - the `fetch_singleton` on the top of the `select_for_update` - so I needed to adjust the syntax here [[ref](https://docs.djangoproject.com/en/5.0/topics/db/managers/#creating-a-manager-with-queryset-methods)].

Comment (GitRon) [reply django_migration_zero/models.py:17]: Ah, now I get it! I like the idea but there is a convention to make a queryset method always return a queryset. That's why I put the `fetch_singleton()` in a manager method.

Do you have an idea on how we could obey by the rule and still get this merged? 🤔

Comment (mad-anne) [reply django_migration_zero/models.py]: Got it! It worked fine on the manager by moving the `select_for_update` inside the `fetch_singleton` (last commit).

**Comment by GitRon** [comment]
Content Review: It seems to work, couldn't break it ✔️

File: django_migration_zero/models.py (hunk 2 above)
Comment (GitRon) [reply django_migration_zero/models.py]: Awesome ❤️

**Review by GitRon — APPROVED** [review]
Amazing work, again! Thx for all your time and effort!

**Comment by GitRon** [comment]
New release: https://pypi.org/project/django_migration_zero/

**Comment by mad-anne** [comment]
Awesome, thank you for the quick release @GitRon!

**Comment by GitRon** [comment]
> Awesome, thank you for the quick release @GitRon!

No worries! Thx for all the great contributions! I hope it works well for you 🙂

**Pull request merged**